runtime/libffi/z/sysvz64.s

```diff
@@ -432,51 +480,360 @@ DARGF    DS 0H                l_DOUBLE in arg area
          LA  14,4             We reached max fprs
          B CONT               Next parameter
 
-*If we have spare gprs, pass up to 12bytes
-*in GPRs. 
-*TODO: Store left over struct to argument area
-*
+*If we have spare gprs, pass up to 24 bytes in GPRs. 
 STRCT    DS 0H
-         L   WRKREG,(2176+(((DSASZ+31)/32)*32)+24)(,4)
-         CFI WRKREG,12
-         BL STRCT2
-         B CONT
-STRCT2   DS 0H
-         LA 15,STRCTS
-         LR WRKREG,GPX
-         SLL WRKREG,2
-         L  15,0(WRKREG,15)
-         BR 15
+         LG  11,0(10,5)
+         LG 15,0(,11)          type->size
+*todo NULL check?
 
-BYTE4    DS 0H
-         L 1,0(6,STKREG)
-         AHI GPX,1
-         B CONT
+*check if first element is float or double 
+STFPCHK  DS 0H  
+         LG 15,16(,11)          type->elements
+         LG 15,0(,15)           type->elements[0]
+         LH 15,10(,15)          type->elements[0]->type
+         CFI 15,11              is it a float?
+         BE STFPCHKF
+         CFI 15,1               is it a double?
+         BE STFPCHKD
+         B STRCT2
 
-BYTE8    DS 0H
-         L 1,0(6,STKREG)
-         L 2,4(6,STKREG)
-         AHI GPX,2
-         AHI 6,4
-         B CONT
+*first arg is a float
+STFPCHKF DS 0H
+         LG 15,16(,11)          type->elements
+         LA 15,8(,15)           type->elements[0]
+         LG 15,0(,15)
+         LH 15,10(,15)          type->elements[0]->type
+         CFI 15,11              is it a float?
+         BE STFPCHKF2
+         B STRCT2
+
+*first arg is a double
+STFPCHKD DS 0H
+         LG 15,16(,11)          type->elements
+         LA 15,8(,15)           type->elements[1]
+         LG 15,0(,15)
+         LH 15,10(,15)          type->elements[1]->type
+         CFI 15,1               is it a double?
+         BE STFPCHKD2
+         B STRCT2
+
+
+*check if there are more elements
+*if not we have a special case
+*I think this is to handle complex types
```

Comment:
It is correct handling of complex types - May be we should have more assertive comment here?